### DISCUSSION TITLE
Error when interacting with FundMe contract with Web3 ValueError: {'code': -32016, 'data': 'Reverted', 'message': 'The execution failed due to an exception.'}

### DISCUSSION THREAD

**Discussion by ptn77** [smartcontractkit/full-blockchain-solidity-course-py#1367] [Q&A] [answered]
I deployed my Fund_me smart contract on Kovan testnet and I tried to call the withdraw function on it using web3, but I'm getting the error ValueError: {'code': -32016, 'data': 'Reverted', 'message': 'The execution failed due to an exception.'}. How can I troubleshoot this or get a more detailed error of what's wrong? Here's the contract withdraw method:

```Solidity
 modifier onlyOwner() {
    require(msg.sender == owner);
    _;
}

function withdraw() public payable onlyOwner {
    msg.sender.transfer(address(this).balance);

    for (
        uint256 funderIndex = 0;
        funderIndex < funders.length;
        funderIndex++
    ) {
        address funder = funders[funderIndex];
        addressToAmountFunded[funder] = 0;
    }
    funders = new address[](0);
}
```
Here's the web3 code to invoke the withdraw method on the contract:

```Python
web3 = Web3(
    Web3.HTTPProvider("https://kovan.infura.io/v3/a7cfd54a6f5147449c622316ea15a4ba")
)
abi = '[{"inputs":[{"internalType":"address","name":"_priceFeed","type":"address"}],"stateMutability":"nonpayable","type":"constructor"},{"inputs":[{"internalType":"address","name":"","type":"address"}],"name":"addressToAmountFunded","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"fund_Catherine","outputs":[],"stateMutability":"payable","type":"function"},{"inputs":[{"internalType":"uint256","name":"","type":"uint256"}],"name":"funders","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"ethAmount","type":"uint256"}],"name":"getConversionRate","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"getEntranceFee","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"getPrice","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"getVersion","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"owner","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"priceFeed","outputs":[{"internalType":"contract AggregatorV3Interface","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"withdraw","outputs":[],"stateMutability":"payable","type":"function"}]'
addr = "0xAAa9F30a71456aFB436dDF49BaA12952F5116017"
modelcontract = web3.eth.contract(address=addr, abi=abi)
tx = modelcontract.functions.withdraw().buildTransaction()
tx.update({"from": str(account)})
tx.update({"to": addr})
tx.update({"gas": 21000})
tx.update({"gasPrice": 2})
nonce = web3.eth.get_transaction_count(str(account))
print(nonce)
tx.update({"nonce": nonce})
signed_message = web3.eth.account.sign_message(
    tx, private_key=config["wallets"]["from_key"]
tx_hash = web3.eth.send_raw_transaction(signed_message.rawTransaction)
tx_receipt = web3.eth.wait_for_transaction_receipt(tx_hash, 180)
```

**Comment by Noveleader** [comment]
Hey @ptn77 I hope you're deploying the right contract, try changing this ```msg.sender.transfer(address(this).balance);```  to 
```payable(msg.sender).transfer(address(this).balance);```

**Reply by ptn77** [reply]
Hi @Noveleader Thank you for responding. The contract is working because I tried it directly on etherscan and the ether amount gets deposited into my wallet correctly. I think there's an issue with the web3 client code.
Directly on the kovan etherscan: https://kovan.etherscan.io/address/0xaaa9f30a71456afb436ddf49baa12952f5116017#writeContract

**Reply by Noveleader** [reply]
ohh great

**Comment by cromewar** [comment, marked as answer]
Hello @ptn77 
Why your Abi is that big?

**Reply by ptn77** [reply]
@cromewar I have many functions on the contract 😆

**Reply by cromewar** [reply]
Well I reviewed your code and seems everything ok by far, and also you have the contract on kovan already, did you figured out what the problem was?

**Reply by ptn77** [reply]
@cromewar I have not tried it again yet. I will maybe later tonight. Thanks

**Reply by cromewar** [reply]
Please let us know later

**Reply by ptn77** [reply]
Ok, I tried putting in print statements to see where it's failing an it looks like it's failing when it builds the transaction from the contract.withdraw() method:

` web3 = Web3(
        Web3.HTTPProvider("https://kovan.infura.io/v3/a7cfd54a6f5147449c622316ea15a4ba")
    )
    abi = '[{"inputs":[{"internalType":"address","name":"_priceFeed","type":"address"}],"stateMutability":"nonpayable","type":"constructor"},{"inputs":[{"internalType":"address","name":"","type":"address"}],"name":"addressToAmountFunded","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"fund_Catherine","outputs":[],"stateMutability":"payable","type":"function"},{"inputs":[{"internalType":"uint256","name":"","type":"uint256"}],"name":"funders","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"ethAmount","type":"uint256"}],"name":"getConversionRate","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"getEntranceFee","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"getPrice","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"getVersion","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"owner","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"priceFeed","outputs":[{"internalType":"contract AggregatorV3Interface","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"withdraw","outputs":[],"stateMutability":"payable","type":"function"}]'
    addr = "0xAAa9F30a71456aFB436dDF49BaA12952F5116017"

    modelcontract = web3.eth.contract(address=addr, abi=abi)
    print("Model contract: " + str(modelcontract.functions.getEntranceFee().call()))
    nonce = web3.eth.get_transaction_count(str(account))
    print("nonce" + str(nonce))
    
    tx = modelcontract.functions.withdraw().buildTransaction()`

Full error:

    tx = modelcontract.functions.withdraw().buildTransaction()
  File "web3/contract.py", line 1079, in buildTransaction
    return build_transaction_for_function(
  File "web3/contract.py", line 1648, in build_transaction_for_function
    prepared_transaction = fill_transaction_defaults(web3, prepared_transaction)
  File "cytoolz/functoolz.pyx", line 250, in cytoolz.functoolz.curry.__call__
  File "web3/_utils/transactions.py", line 102, in fill_transaction_defaults
    default_val = default_getter(web3, transaction)
  File "web3/_utils/transactions.py", line 66, in <lambda>
    'gas': lambda web3, tx: web3.eth.estimate_gas(tx),
  File "web3/eth.py", line 728, in estimate_gas
    return self._estimate_gas(transaction, block_identifier)
  File "web3/module.py", line 57, in caller
    result = w3.manager.request_blocking(method_str,
  File "web3/manager.py", line 187, in request_blocking
    return self.formatted_response(response,
  File "web3/manager.py", line 168, in formatted_response
    raise ValueError(response["error"])
ValueError: {'code': -32016, 'data': 'Reverted', 'message': 'The execution failed due to an exception.'}

**Reply by cromewar** [reply]
Try sending the gasprice on the transacction

**Reply by ptn77** [reply]
Yes, I tried that and I'm getting a new error now.
Here's the updated code:
```Python
web3 = Web3(
        Web3.HTTPProvider("https://kovan.infura.io/v3/a7cfd54a6f5147449c622316ea15a4ba")
    )
    abi = '[{"inputs":[{"internalType":"address","name":"_priceFeed","type":"address"}],"stateMutability":"nonpayable","type":"constructor"},{"inputs":[{"internalType":"address","name":"","type":"address"}],"name":"addressToAmountFunded","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"fund_Catherine","outputs":[],"stateMutability":"payable","type":"function"},{"inputs":[{"internalType":"uint256","name":"","type":"uint256"}],"name":"funders","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[{"internalType":"uint256","name":"ethAmount","type":"uint256"}],"name":"getConversionRate","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"getEntranceFee","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"getPrice","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"getVersion","outputs":[{"internalType":"uint256","name":"","type":"uint256"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"owner","outputs":[{"internalType":"address","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"priceFeed","outputs":[{"internalType":"contract AggregatorV3Interface","name":"","type":"address"}],"stateMutability":"view","type":"function"},{"inputs":[],"name":"withdraw","outputs":[],"stateMutability":"payable","type":"function"}]'
    addr = "0xAAa9F30a71456aFB436dDF49BaA12952F5116017"

    modelcontract = web3.eth.contract(address=addr, abi=abi)
    print("Model contract: " + str(modelcontract.functions.getEntranceFee().call()))
    nonce = web3.eth.get_transaction_count(str(account))
    print("nonce" + str(nonce))

    tx = modelcontract.functions.withdraw().buildTransaction(
        {
            "chainId": 42,
            "gas": 21000,
            "gasPrice": Web3.toWei(2, "gwei"),
            "from": str(account),
            "nonce": nonce,
        }`

```
Model contract getEntranceFee: 14257525121759265
nonce 47

and here's the new error:

{'value': 0, 'chainId': 42, 'gas': 21000, 'gasPrice': 2000000000, 'from': '0xEB6dF80433b81fD866E67Ef5B13bb9A283cb46B9', 'nonce': 47, 'to': '0xAAa9F30a71456aFB436dDF49BaA12952F5116017', 'data': '0x3ccfd60b'}
  File "brownie/_cli/run.py", line 49, in main
    return_value, frame = run(
  File "brownie/project/scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File "./scripts/fund_and_withdraw.py", line 78, in main
    withdraw()
  File "./scripts/fund_and_withdraw.py", line 63, in withdraw
    signed_message = web3.eth.account.sign_message(
  File "eth_utils/decorators.py", line 18, in _wrapper
    return self.method(obj, *args, **kwargs)
  File "eth_account/account.py", line 558, in sign_message
    message_hash = _hash_eip191_message(signable_message)
  File "eth_account/messages.py", line 63, in _hash_eip191_message
    version = signable_message.version
AttributeError: 'dict' object has no attribute 'version'

**Reply by cromewar** [reply]
This one is tricky as the eth_account actually has a version type
![image](https://user-images.githubusercontent.com/20199180/161748385-7cae01ee-d866-4717-8a20-0dea753ab4ed.png)
I think this could be due to a indentation error on your code, but not sure, I'll try to do more research on this.

**Reply by ptn77** [reply]
I think you are right about the indentation. I just re-wrote the code to remove indents:

key = config["wallets"]["from_key"]
# sign the transaction:
signed_message = w3.eth.account.sign_transaction(tx, private_key=key)

Now it's working and it's just complaining about running out of gas now.

thanks for your help

**Reply by cromewar** [reply]
I'm glad is solved now.

**Comment by ptn77** [comment]
I have many functions 😆
